Update to EPAS v0.3 and Event Order Validation

Nfrederiksen opened the pull request:
PR resolves #10

New Validation rules now consider unknown event names, and if certain events, namely (`init`, `loading`, `loaded`, `error`, `stopped`), occur more than once in a session.

It also handles cases for when a `playing` or `paused` event does not have a `paused` or `playing` event preceding it.
There can never be more `paused` events than there have been `playing` events.

In general, the validation logic will base the validity of an event on the name of the previous event in the ordered list (not counting `warning`, `heartbeat`, or `metadata`)

New unit tests that test the validation for different event sequences.  

EDIT: Validation logic has been enhanced. To see a quick overview of how it validates I'd suggest looking at the unit test as I tried to cover as many unique cases as possible.

Nfrederiksen committed:
chore: update event validation order for epas 0.3

Incorperated metadata event in validation order logic, and also
made the validation more strict, following the flowchart in epas spec 0.3

Nfrederiksen committed:
chore: update validation and unittests

Nfrederiksen committed:
chore: add logic for invalidating multiple
occurences of specific events, and update tests

Nfrederiksen committed:
tests: rename validatedEvents to expectedValidatedEvents

Nfrederiksen committed:
chore: clean up logs

oscnord reviewed (approved):
LGTM

martinstark commented on lib/Validator.ts:
```diff
@@ -44,24 +77,56 @@ export default class Validator {
             eventsList[i + 1]['valid'] = true;
           }
         }
+        // Check if 'init' is not first
```
Doesn't that already happen at like 84?

martinstark commented on lib/Validator.ts:
```diff
@@ -44,24 +77,56 @@ export default class Validator {
     try {
-      eventsList[0]['valid'] = true;
+      if (eventsList[0]['event'] !== 'init') {
+        eventsList[0]['valid'] = false;
```
The code is currently mutating the incoming array, which is not good, since it can unpredictably affect unrelated code making use of this package, if it holds references to the original eventsList array.

Non mutating code could (for example) be achieved by mapping the incoming array to a new array of validated events, instead of operating directly on the incoming array: `const validatedEvents = eventsList.map((event: TPlayerAnalyticsEvent) => ValidatedEvent)`.

martinstark commented:
A general suggestion: I think it would make sense if the validator threw an error immediately when it encounters an invalid event, instead of validating the full array and then returning it. 

There is should be no point continuing in validating events if one event is faulty, because the rest of the event sequence will by default be invalid and impossible to validate if once of the preceeding events were out of order.

Nfrederiksen committed:
chore: import epas spec lib

Nfrederiksen committed:
chore: update handlers and type defs in helpers

Nfrederiksen committed:
chore: update ts intefaces

Nfrederiksen committed:
chore: refactor validation logic

new logic covers more cases, namly special rules for playing and paused
events, and also handle cases where warning metadata and heartbeat may
be placed inbetween *ing events

Nfrederiksen committed:
tests: update and make better coverage for...

 some valid and invalid sequences

Nfrederiksen committed:
chore: update readme and fastify script

Nfrederiksen committed:
chore: rename var

Nfrederiksen committed:
chore: update gh actions

Nfrederiksen committed:
chore: clean up

Nfrederiksen committed:
fix typo

Nfrederiksen committed:
chore: default on port 3000

Nfrederiksen committed:
chore: more clean up

slowmove commented on lib/Validator.ts:
```diff
@@ -1,30 +1,65 @@
 
   constructor(logger: winston.Logger) {
     this.logger = logger;
+    this.prevPausedOrPlaying = 'none';
```
Maybe `none` and the following statuses here should rather be enums?

slowmove commented on lib/Validator.ts:
```diff
@@ -1,30 +1,65 @@
   events: any;
+  singleAppearanceEvents: {};
+  output: ValidatorOutput;
+  prevPausedOrPlaying: string;
```
Type as enum or which strings that are valid?

Nfrederiksen committed:
chore: fix enum type for playing paused rule

slowmove commented on lib/Validator.ts:
```diff
@@ -185,22 +193,22 @@ export default class Validator {
     if (event === 'playing') {
-      if (this.prevPausedOrPlaying === 'paused' || this.prevPausedOrPlaying === 'none') {
-        this.prevPausedOrPlaying = 'playing';
+      if (this.prevPlayPauseStatus === PlayPauseStatus.PAUSED || this.prevPlayPauseStatus === PlayPauseStatus.NONE) {
```
When multiple comparisons against the same variable I personally prefer the pattern
`[PlayPauseStatus.PAUSED, PlayPauseStatus.NONE].includes(this.prevPlayPauseStatus)`
Though that might just be my opinion

Nfrederiksen replied:
I like this

slowmove reviewed (approved):
Minor opinion

Nfrederiksen committed:
chore: minor comparison clean up

Nfrederiksen committed:
chore: clean up
